File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/domain/bingo/BingoRoomGame.java
package systems.zlink.samples.bingo.server.play.domain.bingo;

import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

public final class BingoRoomGame {
    private static final String WaitingForPlayers = "WaitingForPlayers";
    private static final String Running = "Running";
    private static final String Finished = "Finished";

    private final String roomId;
    private final BingoRoomModels.BingoRoomSettings settings;
    private final List<BingoRoomModels.RoomPlayer> players = new ArrayList<>();
    private final Queue<Integer> drawDeck = new ArrayDeque<>();
    private final List<Integer> drawnNumbers = new ArrayList<>();
    private final List<String> winners = new ArrayList<>();
    private String status = WaitingForPlayers;

    public BingoRoomGame(String roomId, BingoRoomModels.BingoRoomSettings settings) {
        this.roomId = roomId;
        this.settings = settings;
        for (int number = 1; number <= settings.maxDrawNumber(); number++) {
            drawDeck.add(number);
        }
    }

    public static BingoRoomGame restore(
            String roomId,
            BingoRoomModels.BingoRoomSettings settings,
            Messages.BingoRoomState state) {
        BingoRoomGame game = new BingoRoomGame(roomId, settings);
        game.status = state.getStatus();
        game.drawnNumbers.addAll(state.getDrawnNumbersList());
        game.drawDeck.removeAll(state.getDrawnNumbersList());
        game.winners.addAll(state.getWinnersList());
        for (Messages.BingoPlayerState player : state.getPlayersList()) {
            BingoCard card =
                    player.getCardCount() == 0
                            ? null
                            : BingoCard.restore(player.getCardList(), player.getMarksList());
            game.players.add(
                    new BingoRoomModels.RoomPlayer(
                            player.getActorId(),
                            player.getDisplayName(),
                            player.getSeat(),
                            card,
                            player.getWins(),
                            player.getLosses()));
        }
        return game;
    }

    public Change join(String actorId, String displayName, int wins, int losses) {
        BingoRoomModels.RoomPlayer existing = player(actorId);
        if (existing != null) {
            return new Change(snapshot(), List.of(), false);
        }
        if (!status.equals(WaitingForPlayers) || players.size() >= settings.requiredPlayers()) {
            throw new IllegalStateException("Room " + roomId + " cannot accept more players.");
        }
        BingoRoomModels.RoomPlayer joined =
                new BingoRoomModels.RoomPlayer(
                        actorId, displayName, players.size(), null, wins, losses);
        players.add(joined);
        Messages.BingoRoomState joinedState = snapshot();
        ArrayList<BingoRoomModels.RoomEvent> events =
                new ArrayList<>(playerJoinedEvents(joined, joinedState));
        boolean started = false;
        if (players.size() == settings.requiredPlayers()) {
            status = Running;
            started = true;
            Messages.BingoRoomState startedState = snapshot();
            events.addAll(eventsForAll(BingoRoomModels.EventKind.GAME_STARTED, startedState));
            return new Change(startedState, events, true);
        }
        return new Change(joinedState, events, false);
    }

    public boolean canAcceptPlayer() {
        return status.equals(WaitingForPlayers) && players.size() < settings.requiredPlayers();
    }

    public Messages.BingoRoomState previewJoin(String actorId, String displayName) {
        BingoRoomModels.RoomPlayer existing = player(actorId);
        if (existing != null) {
            return snapshot();
        }
        if (!status.equals(WaitingForPlayers) || players.size() >= settings.requiredPlayers()) {
            throw new IllegalStateException("Room " + roomId + " cannot accept more players.");
        }
        ArrayList<BingoRoomModels.RoomPlayer> previewPlayers = new ArrayList<>(players);
        previewPlayers.add(
                new BingoRoomModels.RoomPlayer(
                        actorId, displayName, previewPlayers.size(), null, 0, 0));
        String previewStatus =
                previewPlayers.size() == settings.requiredPlayers() ? Running : status;
        return snapshot(previewPlayers, previewStatus);
    }

    public Change submitCard(String actorId, List<Integer> submittedCard) {
        int index = playerIndex(actorId);
        if (!status.equals(Running)) {
            throw new IllegalStateException(
                    "bingo card can be submitted only after the room starts");
        }
        BingoRoomModels.RoomPlayer current = players.get(index);
        if (current.card() != null) {
            throw new IllegalStateException("bingo card was already submitted");
        }
        players.set(
                index,
                new BingoRoomModels.RoomPlayer(
                        current.actorId(),
                        current.displayName(),
                        current.seat(),
                        BingoCard.from(submittedCard),
                        current.wins(),
                        current.losses()));
        return new Change(snapshot(), List.of(), allCardsSubmitted());
    }

    public Change drawNext() {
        if (!status.equals(Running) || !allCardsSubmitted() || drawDeck.isEmpty()) {
            return new Change(snapshot(), List.of(), allCardsSubmitted());
        }

        int number = drawDeck.remove();
        drawnNumbers.add(number);
        ArrayList<String> newlyCompleted = new ArrayList<>();
        for (BingoRoomModels.RoomPlayer player : players) {
            int before = player.card().completedLines();
            player.card().markDrawnNumber(number);
            if (player.card().completedLines() > before) {
                newlyCompleted.add(player.actorId());
            }
        }

        if (!newlyCompleted.isEmpty()) {
            status = Finished;
            winners.addAll(newlyCompleted);
        }

        Messages.BingoRoomState state = snapshot();
        ArrayList<BingoRoomModels.RoomEvent> events =
                new ArrayList<>(numberDrawnEvents(state, number));
        events.addAll(
                eventsForAll(
                        status.equals(Finished)
                                ? BingoRoomModels.EventKind.GAME_ENDED
                                : BingoRoomModels.EventKind.STATE,
                        state));
        return new Change(state, events, false);
    }

    public Messages.BingoRoomState snapshot() {
        return snapshot(players, status);
    }

    private Messages.BingoRoomState snapshot(
            List<BingoRoomModels.RoomPlayer> snapshotPlayers, String snapshotStatus) {
        String hostActorId = snapshotPlayers.isEmpty() ? "" : snapshotPlayers.getFirst().actorId();
        Integer lastDrawn = drawnNumbers.isEmpty() ? null : drawnNumbers.getLast();
        return BingoMessages.bingoRoomState(
                roomId,
                snapshotStatus,
                hostActorId,
                false,
                drawnNumbers.size(),
                lastDrawn,
                List.copyOf(drawnNumbers),
                snapshotPlayers.stream().map(player -> player.toState(hostActorId)).toList(),
                List.copyOf(winners));
    }

    private boolean allCardsSubmitted() {
        return players.size() == settings.requiredPlayers()
                && players.stream().allMatch(player -> player.card() != null);
    }

    private BingoRoomModels.RoomPlayer player(String actorId) {
        return players.stream()
                .filter(player -> player.actorId().equals(actorId))
                .findFirst()
                .orElse(null);
    }

    private int playerIndex(String actorId) {
        for (int i = 0; i < players.size(); i++) {
            if (players.get(i).actorId().equals(actorId)) {
                return i;
            }
        }
        throw new IllegalStateException("player has not joined the bingo room");
    }

    private List<BingoRoomModels.RoomEvent> playerJoinedEvents(
            BingoRoomModels.RoomPlayer joined, Messages.BingoRoomState state) {
        return players.stream()
                .filter(player -> !player.actorId().equals(joined.actorId()))
                .map(
                        player ->
                                new BingoRoomModels.RoomEvent(
                                        BingoRoomModels.EventKind.PLAYER_JOINED,
                                        player.actorId(),
                                        state,
                                        joined.actorId(),
                                        joined.displayName(),
                                        joined.seat(),
                                        joined.seat() == 0,
                                        0))
                .toList();
    }

    private List<BingoRoomModels.RoomEvent> numberDrawnEvents(
            Messages.BingoRoomState state, int number) {
        return players.stream()
                .map(
                        player ->
                                new BingoRoomModels.RoomEvent(
                                        BingoRoomModels.EventKind.NUMBER_DRAWN,
                                        player.actorId(),
                                        state,
                                        null,
                                        null,
                                        -1,
                                        false,
                                        number))
                .toList();
    }

    private List<BingoRoomModels.RoomEvent> eventsForAll(
            BingoRoomModels.EventKind kind, Messages.BingoRoomState state) {
        return players.stream()
                .map(
                        player ->
                                new BingoRoomModels.RoomEvent(
                                        kind, player.actorId(), state, null, null, -1, false, 0))
                .toList();
    }

    public record Change(
            Messages.BingoRoomState state,
            List<BingoRoomModels.RoomEvent> events,
            boolean drawReady) {}
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/domain/bingo/BingoRoomGame.kt
package systems.zlink.samples.kotlin.bingo.server.play.domain.bingo

import java.util.ArrayDeque
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomState
import systems.zlink.samples.kotlin.bingo.shared.contracts.card
import systems.zlink.samples.kotlin.bingo.shared.contracts.drawnNumbers
import systems.zlink.samples.kotlin.bingo.shared.contracts.marks
import systems.zlink.samples.kotlin.bingo.shared.contracts.players
import systems.zlink.samples.kotlin.bingo.shared.contracts.winners

class BingoRoomGame(private val roomId: String, private val settings: BingoRoomSettings) {
    private val players = mutableListOf<BingoRoomPlayer>()
    private val drawDeck = ArrayDeque<Int>()
    private val drawnNumbers = mutableListOf<Int>()
    private val winners = mutableListOf<String>()
    private var status: String = WaitingForPlayers

    init {
        for (number in 1..settings.maxDrawNumber) {
            drawDeck.add(number)
        }
    }

    fun join(actorId: String, displayName: String, wins: Int, losses: Int): Change {
        val existing = player(actorId)
        if (existing != null) {
            return Change(snapshot(), emptyList(), false)
        }
        if (status != WaitingForPlayers || players.size >= settings.requiredPlayers) {
            throw IllegalStateException("Room $roomId cannot accept more players.")
        }

        val joined = BingoRoomPlayer(actorId, displayName, players.size, null, wins, losses)
        players += joined
        val joinedState = snapshot()
        val events = playerJoinedEvents(joined, joinedState).toMutableList()
        if (players.size == settings.requiredPlayers) {
            status = Running
            val startedState = snapshot()
            events += eventsForAll(BingoRoomEventKind.GAME_STARTED, startedState)
            return Change(startedState, events, true)
        }
        return Change(joinedState, events, false)
    }

    fun canAcceptPlayer(): Boolean =
        status == WaitingForPlayers && players.size < settings.requiredPlayers

    fun previewJoin(actorId: String, displayName: String): BingoRoomState {
        val existing = player(actorId)
        if (existing != null) {
            return snapshot()
        }
        if (status != WaitingForPlayers || players.size >= settings.requiredPlayers) {
            throw IllegalStateException("Room $roomId cannot accept more players.")
        }
        val previewPlayers =
            players + BingoRoomPlayer(actorId, displayName, players.size, null, 0, 0)
        val previewStatus = if (previewPlayers.size == settings.requiredPlayers) Running else status
        return snapshot(previewPlayers, previewStatus)
    }

    fun submitCard(actorId: String, submittedCard: List<Int>): Change {
        val index = playerIndex(actorId)
        if (status != Running) {
            throw IllegalStateException("bingo card can be submitted only after the room starts")
        }
        val current = players[index]
        if (current.card != null) {
            throw IllegalStateException("bingo card was already submitted")
        }
        players[index] =
            BingoRoomPlayer(
                current.actorId,
                current.displayName,
                current.seat,
                BingoCard.from(submittedCard),
                current.wins,
                current.losses,
            )
        return Change(snapshot(), emptyList(), allCardsSubmitted())
    }

    fun drawNext(): Change {
        if (status != Running || !allCardsSubmitted() || drawDeck.isEmpty()) {
            return Change(snapshot(), emptyList(), allCardsSubmitted())
        }

        val number = drawDeck.remove()
        drawnNumbers += number
        val newlyCompleted = mutableListOf<String>()
        for (player in players) {
            val card = player.card ?: continue
            val before = card.completedLines()
            card.markDrawnNumber(number)
            if (card.completedLines() > before) {
                newlyCompleted += player.actorId
            }
        }

        if (newlyCompleted.isNotEmpty()) {
            status = Finished
            winners += newlyCompleted
        }

        val state = snapshot()
        val events = numberDrawnEvents(state, number).toMutableList()
        events +=
            eventsForAll(
                if (status == Finished) BingoRoomEventKind.GAME_ENDED else BingoRoomEventKind.STATE,
                state,
            )
        return Change(state, events, false)
    }

    fun snapshot(): BingoRoomState = snapshot(players, status)

    private fun snapshot(
        snapshotPlayers: List<BingoRoomPlayer>,
        snapshotStatus: String,
    ): BingoRoomState {
        val hostActorId = snapshotPlayers.firstOrNull()?.actorId ?: ""
        val lastDrawn = drawnNumbers.lastOrNull()
        return BingoRoomState(
            roomId,
            snapshotStatus,
            hostActorId,
            false,
            drawnNumbers.size,
            lastDrawn,
            drawnNumbers.toList(),
            snapshotPlayers.map { it.toState(hostActorId) },
            winners.toList(),
        )
    }

    private fun allCardsSubmitted(): Boolean =
        players.size == settings.requiredPlayers && players.all { it.card != null }

    private fun player(actorId: String): BingoRoomPlayer? =
        players.firstOrNull { it.actorId == actorId }

    private fun playerIndex(actorId: String): Int {
        players.forEachIndexed { index, player ->
            if (player.actorId == actorId) {
                return index
            }
        }
        throw IllegalStateException("player has not joined the bingo room")
    }

    private fun playerJoinedEvents(
        joined: BingoRoomPlayer,
        state: BingoRoomState,
    ): List<BingoRoomEvent> =
        players.mapNotNull { player ->
            if (player.actorId == joined.actorId) {
                return@mapNotNull null
            }
            BingoRoomEvent(
                BingoRoomEventKind.PLAYER_JOINED,
                player.actorId,
                state,
                joined.actorId,
                joined.displayName,
                joined.seat,
                joined.seat == 0,
                0,
            )
        }

    private fun numberDrawnEvents(state: BingoRoomState, number: Int): List<BingoRoomEvent> =
        players.map { player ->
            BingoRoomEvent(
                BingoRoomEventKind.NUMBER_DRAWN,
                player.actorId,
                state,
                null,
                null,
                -1,
                false,
                number,
            )
        }

    private fun eventsForAll(
        kind: BingoRoomEventKind,
        state: BingoRoomState,
    ): List<BingoRoomEvent> =
        players.map { player ->
            BingoRoomEvent(kind, player.actorId, state, null, null, -1, false, 0)
        }

    data class Change(
        val state: BingoRoomState,
        val events: List<BingoRoomEvent>,
        val drawReady: Boolean,
    )

    companion object {
        const val WaitingForPlayers = "WaitingForPlayers"
        const val Running = "Running"
        const val Finished = "Finished"

        fun restore(
            roomId: String,
            settings: BingoRoomSettings,
            state: BingoRoomState,
        ): BingoRoomGame =
            BingoRoomGame(roomId, settings).also { game ->
                game.status = state.status
                game.drawnNumbers += state.drawnNumbers
                game.drawDeck.removeAll(state.drawnNumbers.toSet())
                game.winners += state.winners
                state.players.forEach { player ->
                    game.players +=
                        BingoRoomPlayer(
                            actorId = player.actorId,
                            displayName = player.displayName,
                            seat = player.seat,
                            card =
                                player.card
                                    .takeIf { it.isNotEmpty() }
                                    ?.let { BingoCard.restore(it, player.marks) },
                            wins = player.wins,
                            losses = player.losses,
                        )
                }
            }
    }
}
